app/classes/content_filter.rb

```diff
@@ -1,96 +1,117 @@
+  # automatically.)
+  #
+  #   Add tests, e.g., to TestUserContentFilter
+  #   Supplement fixtures as needed by added tests
+  #   Add a filter definition below
+  # * Add a new checkbox_val method in User
+  #   In /config/locales/en.txt define text to be displayed before and next to
+  #     check boxes, using prefs_obs_filters_has_images: as a model.
+  # * Supplement _prefs_filters.html.erb as needed
+  # * Add filter to AccountController#prefs_types & #update_content_filter
+  #   For Observation filter, supplement Query::Initializers::ObservationFilters
+  # * Supplement ApplicationController#show_index_of_objects as needed
+  #   To filter another object, create a new initializer and include in
+  #     appropriate searches.
+  #
+  # To be able to override the new filter in Advanced Searches, at least:
+  #   Add tests, e.g., to TestAdvancedSearchFilters
+  #   Supplement fixtures as needed by added tests
+  #   In /config/locales/en.txt define text to be displayed before and next to
+  #     radio boxes, using advanced_search_filter_has_images: as a model.
+  # * Supplement _advanced_search_filters.html.erb as needed
+  # * Supplement ObservationController#advanced_search_form as needed.
+  # * Supplement ApplicationController#show_index_of_objects as needed.
+  # * Supplement Query::RssLogBase as needed.
+  #
+  # To also add it as a Pattern Search
+  #   Supplement PatternSearchTest
+  #   Supplement classes/pattern_search/observation.rb or -- if it's not an
+  #     Observation search -- add a new pattern_search file and class.
+  #
+  # There are probably other steps/files I've forgotten. JDC 2016-09-15
 
-  def has_images
-    {
-      sym:          :has_images,
-      model:        Observation,
-      checkbox:     :has_images_checkbox,
-      checked_val:  "NOT NULL",
-      off_val:      "off",
-      on_vals:      ["NOT NULL", "NULL"],
-      sql_cond:     "observations.thumb_image_id IS #{params[:has_images]}"
-    }
-  end
+  attr_accessor :sym
+  attr_accessor :model
+  attr_accessor :checkbox
+  attr_accessor :on_vals
+  attr_accessor :checked_val
+  attr_accessor :off_val
 
-  def has_specimen
-    {
-      sym:          :has_specimen,
-      model:        Observation,
-      checkbox:     :has_specimen_checkbox,
-      checked_val:  "TRUE",
-      off_val:      "off",
-      on_vals:      ["TRUE", "FALSE"],
-      sql_cond:     "observations.specimen IS #{params[:has_specimen]}"
-   }
+  def initialize(opts)
+    @sym          = opts[:sym]
+    @model        = opts[:model]
+    @checkbox     = opts[:checkbox]
+    @on_vals      = opts[:on_vals]
+    @checked_val  = opts[:checked_val]
+    @off_val      = opts[:off_val]
   end
 
-  def filters
-    [has_images, has_specimen]
+  # Array of all available filters.
+  def self.all
+    @@filters ||= [
+      ContentFilter.new(
+        sym:          :has_images,
+        model:        Observation,
+        checkbox:     :has_images_checkbox,
```

Comment:
Yes.  
When I wrote all this code, I was thinking like a procedural programmer rather than an OOP program -- which I am not.  I love the ContentFilters as a class; couldn't figure out how to do that on my own.
Also note @pellaea email comment (pasted here for completeness):

> d) It is possible to have the User#has_specimen_checkbox and User#has_image_checkbox methods created dynamically based on ContentFilter using a construct like this:
```ruby
  ContentFilter.all.each do |fltr|
    next unless fltr.checkbox.present?
    define_method(fltr.checkbox) do
      content_filter[fltr.sym] == fltr.checked_val ? 1 : 0
    end
  end
```